File: jinhyeok/04_30/C++/3863.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int main() {
  // ios_base::sync_with_stdio와 cin.tie(NULL)를 사용하여 입출력 속도를 향상시킴.
  ios_base::sync_with_stdio(false);
  cin.tie(NULL);

  int n, m;

  // n, m을 입력받고 두 변수가 0이 아닌 동안 반복.
  while (cin >> n >> m && n != 0 && m != 0) {
    // 전화 정보를 저장하는 벡터.
    vector<pair<long long, long long>> calls(n);

    // 각 전화 통화의 정보를 입력받고, calls 벡터에 시작 시간과 끝나는 시간을 저장.
    for (int i = 0; i < n; i++) {
      long long src, dest, start, dur;
      cin >> src >> dest >> start >> dur;
      calls[i] = make_pair(start, start + dur);
    }

    for (int i = 0; i < m; i++) {
      long long start, dur; // 구간의 시작 시간과 지속 시간.
      cin >> start >> dur;
      long long end = start + dur; // 구간의 끝나는 시간 계산.
      int count = 0; // 구간에 포함되는 전화 통화의 수를 저장할 변수를 초기화 시킴.
      for (const auto& call : calls) { // 각 전화에 대해 반복.
        long long cStart = call.first; // 전화 통화의 시작 시간.
        long long cEnd = call.second; // 전화 통화의 끝나는 시간.
        if (min(end, cEnd) - max(cStart, start) >= 1) { // 구간에 포함되는 전화 통화의 조건을 확인하고, 포함되는 경우 count++
          count++;
        }
      }
      cout << count << endl;
    }
  }
  return 0;
}

File: jinhyeok/04_30/Js/3863.js
// 표준 입력을 읽을 수 있도록 설정.
process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputLines = []; // 입력된 줄을 저장할 변수.
let currentLine = 0; // 현재 처리 중인 줄의 인덱스 추적.

// 입력 데이터를 받아 inputLines 배열에 저장하고, main() 함수를 호출.
process.stdin.on('data', inputStdin => {
    inputLines = inputStdin.split('\n');
    main();
});

// inputLines에서 한 줄씩 읽어오는 함수.
function readLine() {
    return inputLines[currentLine++];
}

function main() {
    while (true) {
        // n, m 입력받기.
        const [n, m] = readLine().split(' ').map(Number);
        if (n === 0 && m === 0) { // 모두 0인 경우만 종료.
            break;
        }

        const calls = []; // 전화 통화 정보를 저장할 배열 초기화.
        // 전화 통화 시작 시간과 끝 시간을 저장하는 배열.
        for (let i = 0; i < n; i++) {
            const [src, dest, start, dur] = readLine().split(' ').map(Number);
            calls.push([start, start + dur]);
        }

        const results = [];
        for (let i = 0; i < m; i++) {
            // 구간의 시작 시간과 지속 시간을 입력받음.
            const [start, dur] = readLine().split(' ').map(Number);
            const end = start + dur; // 구간의 끝나는 시간 계산.
            // 구간에 포함되는 전화 통화의 수를 계산.
            const count = calls.reduce((acc, [cStart, cEnd]) => {
                // 구간에 포함되는 전화 통화의 조건을 확인, 포함되는 경우 +1 그렇지 않으면 +0
                return acc + (Math.min(end, cEnd) - Math.max(cStart, start) >= 1 ? 1 : 0);
            }, 0);
            // 구간의 결과를 results 배열에 저장.
            results.push(count);
        }

        console.log(results.join('\n'));
    }
}